### DISCUSSION TITLE
How to customize the error message of `z.coerce.date()`?

### DISCUSSION THREAD

**Discussion by tordans** [colinhacks/zod#1851] [Q&A] [answered]
When I use `z.date()` I can customize the error message like …

```ts
z.date({
  required_error: "Date is required.",
  invalid_type_error: "Wrong date format.",
}),
```

However, once I use `z.coerce.date()`  the error messages are ignored (or "not reached").

For example, `z.coerce.date().parse('')` returns…

```ts
Uncaught ZodError: [
  {
    "code": "invalid_date",
    "path": [],
    "message": "Invalid date"
  }
]
```

I assume that ZodError gets this message from `new Date('')`, which also return `Invalid date`.

However, since I use react-hook-form with a zod schema, I "have to" provide the error messages via zod (or start doing some hacky workarounds).

As far as I see, right now there is no way to change this behavior (eg. `coerce` does not accept an error message).

Is this considered a Bug or a feature in Zod? And how would you recommend working around it?

---

A few thoughts on how to improve the DX:
* the error message from `.date` could overwrite that by `.coerce`
* `.coerce` could get an optional message to be used
* if possible TS could mark `required_error` and `invalid_type_error` as `never` whenever `.date` is chained on top of `.coerce`, since those message will never work
* the docs could hint at this special behavior

**Comment by JacobWeisenburger** [comment, marked as answer]
Is this what you are looking for?

```ts
const schema = z.coerce.date( {
    errorMap: ( { code }, { defaultError } ) => {
        if ( code == 'invalid_date' ) return { message: 'Wrong date format.' }
        return { message: defaultError }
    }
} )

const result = schema.safeParse( '' )
!result.success && console.log( result.error.issues[ 0 ] )
// { code: 'invalid_date', path: [], message: 'Wrong date format.' }
```

**Reply by tordans** [reply]
@JacobWeisenburger thanks a lot, that is exactly what I needed. Would a PR be welcome to hint at this [in the coerse section](https://zod.dev/?id=coercion-for-primitives)?

**Reply by JacobWeisenburger** [reply]
PRs are always welcome. Let me know when it's ready and I'll review it.

**Reply by lsa-jose-israel** [reply]
Shouldnt the message be overriden if you pass it in the { message } parameter? Do you need to define a whole custom map for every date field?
